node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+{
+	type CreateProposer = Pin<Box<
+		dyn Future<Output = Result<Self::Proposer, Self::Error>> + Send + 'static,
+	>>;
+	type Proposer = Proposer<TxPool, Backend, Client>;
+	type Error = Error;
+
+	fn init(&mut self, parent_header: &Header) -> Self::CreateProposer {
+		// create the inner proposer
+		let proposer = self.inner.init(parent_header).into_inner();
+
+		// clone the overseer handle (lightweight) so it can be moved into the future
+		let mut overseer = self.overseer.clone();
+
+		// we know that this function will be called at least once per proposed block,
+		// because this is where the parent header is supplied. Therefore, we can send
+		// the request for authorship data here.
+		let (sender, receiver) = futures::channel::oneshot::channel();
+		let parent_header_hash = parent_header.hash();
+
+		async move {
+			overseer.send_msg(AllMessages::Provisioner(
+				ProvisionerMessage::RequestInherentData(parent_header_hash, sender),
+			)).await?;
+			Ok(Proposer {
+				inner: proposer?,
+				provisioner_inherent_data: receiver,
+			})
+		}.boxed()
+	}
+}
+
+pub struct Proposer<TxPool: TransactionPool<Block = Block>, Backend, Client> {
+	inner: sc_basic_authorship::Proposer<Backend, Block, Client, TxPool>,
+	provisioner_inherent_data: futures::channel::oneshot::Receiver<ProvisionerInherentData>,
+}
+
+impl<TxPool, Backend, Client> sp_consensus::Proposer<Block> for Proposer<TxPool, Backend, Client>
+where
+	TxPool: 'static + TransactionPool<Block = Block>,
+	Client: 'static
+		+ BlockBuilderProvider<Backend, Block, Client>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Send
+		+ Sync,
+	Client::Api:
+		ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
+	Backend:
+		'static + sc_client_api::Backend<Block, State = sp_api::StateBackendFor<Client, Block>>,
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type Transaction = sc_client_api::TransactionFor<Backend, Block>;
+	type Proposal = Pin<Box<
+		dyn Future<Output = Result<Proposal<Block, sp_api::TransactionFor<Client, Block>>, Error>> + Send,
+	>>;
+	type Error = Error;
+
+	fn propose(
+		self,
+		mut inherent_data: InherentData,
+		inherent_digests: DigestFor<Block>,
+		max_duration: time::Duration,
+		record_proof: RecordProof,
+	) -> Self::Proposal {
+		let mut proposal = async move {
+			let provisioner_inherent_data = self.provisioner_inherent_data.await.map_err(Error::ClosedChannelFromProvisioner)?;
+
+			inherent_data.put_data(
+				inclusion_inherent::INHERENT_IDENTIFIER,
+				&provisioner_inherent_data,
+			)?;
+
+			self.inner
+				.propose(inherent_data, inherent_digests, max_duration, record_proof)
+				.await
+				.map_err(Into::into)
+		}
+		.boxed()
+		.fuse();
+
+		let mut timeout = wasm_timer::Delay::new(PROPOSE_TIMEOUT).fuse();
```

Comment:
My intention with the suggestion of the `PROPOSE_TIMEOUT` was to have a timeout on the specific part of the pipeline that fetches the data from the provisioner. So something like this

```rust
let get_data_from_provisioner = async move {

};

let fallback = wasm_timer::Delay::new(PROPOSE_TIMEOUT).map(|_| empty_inherent_data).fuse();

let inherent_data = get_data_from_provisioner.select(fallback);

// call inner proposer.proposer using inherent data
```

The goal is to author a block even if the overseer/provisioner is not responding. And we should probably log a warning if the fallback is triggered

Reply:
9bae56ed